In [1]:
import selenium
from selenium import webdriver
from bs4 import BeautifulSoup,SoupStrainer
import requests
import pandas as pd
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.common.exceptions import NoSuchElementException
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.action_chains import ActionChains
from selenium.common.exceptions import ElementClickInterceptedException
from selenium.webdriver.support.ui import WebDriverWait
import re
import time

import warnings
warnings.filterwarnings('ignore')

In [2]:
#Sending request to scrap the website
url='https://www.nike.com'
page = requests.get(url)
page

<Response [200]>

In [3]:
#Opening the website in the automated browser
driver = webdriver.Chrome()
page_source = driver.get("https://www.nike.com")

In [4]:
#Maximizing the window
driver.maximize_window()

In [5]:
soup= BeautifulSoup(page.content)

In [6]:
#Sending the input keys
search_input = driver.find_element(By.XPATH,"/html/body/div[3]/div/div[3]/header/div[1]/div[2]/div/div/div[1]/div/div/input")
search_input.send_keys('shoes')

In [7]:
search_button= driver.find_element(By.XPATH,"/html/body/div[3]/div/div[3]/header/div[1]/div[2]/div/div/div[2]/div[2]/div[1]/ul/li[1]/a/span")
search_button.click()

In [8]:
#Creating the empty lists to store the product details
time.sleep(5)
names = []
no_of_colors = []
category = []
price = []
count_of_sizes = []
color1= []
color2= []
color3= []
color4 = []
color5 = []
productcode = []
reviews = []
size =[]
ratings = []
comfort = []
durability = []

In [11]:
#To load all the products in the website
for _ in range(20):
    driver.find_element(By.TAG_NAME,'body').send_keys(Keys.END)
    time.sleep(10)

In [10]:
#Extracting names
for i in driver.find_elements(By.CLASS_NAME,"product-card__title"):
    names.append(i.text)
#Extracting no_of_colors   
for i in driver.find_elements(By.CLASS_NAME,"product-card__count-item"):
    no_of_colors.append(i.text)
#Extracting Category
for i in driver.find_elements(By.CLASS_NAME,"product-card__subtitle"):
    category.append(i.text)
#Extracting price
for i in driver.find_elements(By.CLASS_NAME,"product-card__animation_wrapper"):
    price.append(i.text)

In [18]:
len(names), len(no_of_colors), len(category), len(price)

(981, 981, 981, 981)

In [19]:
driver.execute_script("window.scrollTo(0, 0);")

In [126]:
#Extracting the product details
product_details = driver.find_elements(By.XPATH, '//img[@class="product-card__hero-image css-1fxh5tw"]')

for index in range(len(product_details)):
    try:
        # Re-fetch the product details on each iteration
        product_details = driver.find_elements(By.XPATH, '//img[@class="product-card__hero-image css-1fxh5tw"]')

        # Click on the current product detail
        current_product = product_details[index]
        current_product.click()
        
        # Extract color information
        color_elements = driver.find_elements(By.XPATH, '//*[@id="ColorwayDiv"]/div/div/fieldset/div/div/label/img')
        for i in range(len(color_elements)):
            try:
                color1.append(color_elements[i].get_attribute("alt"))
            except NoSuchElementException:
                color1.append("-")
            if i + 1 < len(color_elements):
                try:
                    color2.append(color_elements[i + 1].get_attribute("alt"))
                except NoSuchElementException:
                    color2.append("-")
                i += 1
            else:
                color2.append("-")
            if i + 1 < len(color_elements):
                try:
                    color3.append(color_elements[i + 1].get_attribute("alt"))
                except NoSuchElementException:
                    color3.append("-")
                i += 1
            else:
                color3.append("-")
            if i + 1 < len(color_elements):
                try:
                    color4.append(color_elements[i + 1].get_attribute("alt"))
                except NoSuchElementException:
                    color4.append("-")
                i += 1
            else:
                color4.append("-")
            if i + 1 < len(color_elements):
                try:
                    color5.append(color_elements[i + 1].get_attribute("alt"))
                except NoSuchElementException:
                    color5.append("-")
                i += 1
            else:
                color5.append("-")
                
        # Extract sizes and count of sizes
        size_elements = driver.find_elements(By.CLASS_NAME, "css-xf3ahq")
        sizes = [size.text for size in size_elements]
        count_of_sizes.append(len(sizes))
        
        driver.execute_script("window.scrollBy(0, 1000)")
        # Extract product code
        productcode_element = driver.find_element(By.CLASS_NAME, "description-preview__style-color")
        productcode.append(productcode_element.text.strip().split(':')[1])

        # Extract reviews
        reviews_element = driver.find_element(By.CLASS_NAME, "css-rptnlm")
        reviews.append(reviews_element.text.split('(')[1].strip(')'))
        
      
        #Extract Comfort
        button = driver.find_element(By.XPATH, '/html/body/div[4]/div/div/div[2]/div/div[4]/div[2]/div[2]/div/div/span/div/button')
        button.click()
        try:
            comfort_elements =  driver.find_element(By.XPATH, '/html/body/div[5]/div/div[2]/div/div/div[2]/div/div/ul[1]')
            comfort.append(comfort_elements.text)
        except Exception as e:
            comfort.append("-")
        close_button = driver.find_element(By.XPATH, '/html/body/div[5]/div/div[2]/div/button')
        close_button.click()
        
        # Extract ratings
        try:
            button = driver.find_element(By.XPATH,'//summary[@class=" css-rptnlm"]')
            button.click()
            try:
                ratings_element = driver.find_element(By.XPATH, '//p[@class="d-sm-ib pl4-sm"]')
                raw_rating = ratings_element.text.strip(" stars")
                rating = raw_rating.split()[0]
                ratings.append(rating)
            except NoSuchElementException:
                ratings.append("-")
        except NoSuchElementException:
            ratings.append("-")

        # Navigate back to the product list page
        driver.back()

    except Exception as e:
        print(f"Error processing element: {str(e)}")

In [87]:
driver.execute_script('return document.body.scrollHeight')

10069

In [88]:
driver.execute_script("window.scrollBy(0, 3500)")

In [128]:
#closing the browser
driver.quit()

In [147]:
table1 = pd.DataFrame(list(zip(names, no_of_colors, category,price)),
                  columns=['Name', 'No_of_Colors', 'Category', 'Price'])

In [170]:
csv_file_path = 'C:/Users/vaddi/.jupyter/table1.csv'

# Write the DataFrame to the CSV file
table1.to_csv(csv_file_path, index=True)

# Display the CSV file path
csv_file_path

'C:/Users/vaddi/.jupyter/table1.csv'

In [171]:
table2= pd.DataFrame(list(zip(count_of_sizes, color1,color2,color3,color4,color5,productcode)),columns=['Sizes','color1','color2','color3','color4','color5','product code'])

In [172]:
csv_file_path = 'C:/Users/vaddi/.jupyter/table2.csv'

# Write the DataFrame to the CSV file
table2.to_csv(csv_file_path, index=True)

# Display the CSV file path
csv_file_path

'C:/Users/vaddi/.jupyter/table2.csv'

In [166]:
for _ in range(635):
    size.append("-")
for _ in range(635):
    durability.append("-")

In [167]:
table3= pd.DataFrame(list(zip(reviews,size,comfort,durability,ratings)),columns=['Reviews','Size','Comfort','Durability','Ratings'])

In [168]:
csv_file_path = 'C:/Users/vaddi/.jupyter/table3.csv'

# Write the DataFrame to the CSV file
table3.to_csv(csv_file_path, index=True)

# Display the CSV file path
csv_file_path

'C:/Users/vaddi/.jupyter/table3.csv'

C:\Users\vaddi\.jupyter

In [87]:
import selenium
from selenium import webdriver
from bs4 import BeautifulSoup,SoupStrainer
from bs4 import BeautifulSoup
import requests
import pandas as pd
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.common.exceptions import NoSuchElementException
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.action_chains import ActionChains
from selenium.common.exceptions import ElementClickInterceptedException
from selenium.common.exceptions import StaleElementReferenceException, TimeoutException

import re
import time

import warnings
warnings.filterwarnings('ignore')

In [130]:
#opening the website in the automated browser
url='https://www.adidas.com'
driver = webdriver.Chrome()
page_source = driver.get(url)

In [131]:
#maximizing the window
driver.maximize_window()

In [136]:
#Sending the input keys in the search bar
search_input = driver.find_element(By.XPATH,'/html/body/div[1]/div[2]/div[2]/div/header/div[2]/div/div[2]/div/input')
search_input.send_keys("Shoes")

In [137]:
search = driver.find_element(By.XPATH,'/html/body/div[1]/div[2]/div[2]/div/header/div[2]/div/div[2]/div/div[2]/div/div[1]/div[3]/a')
search.click()

In [6]:
#Creating empty lists to store the product details
time.sleep(5)
names = []
no_of_colors = []
category = []
sale_price = []
price = []
count_of_sizes = []
color1= []
color2= []
color3= []
color4 = []
color5 = []
productcode = []
reviews = []
size =[]
ratings = []
comfort = []
durability = []

In [7]:
while True:
    try:
        #Extracting names
        n = WebDriverWait(driver, 10).until(EC.presence_of_all_elements_located((By.CLASS_NAME, "glass-product-card__title")))
        for i in n:
            names.append(i.text)

        #Extracting Category
        c = WebDriverWait(driver, 10).until(EC.presence_of_all_elements_located((By.CLASS_NAME, "glass-product-card__category")))
        for i in c:
            category.append(i.text)

        #Extracting no_of_colors
        col = WebDriverWait(driver, 10).until(EC.presence_of_all_elements_located((By.CLASS_NAME, "dark-grey___6ysQv")))
        if col:
            for i in col:
                no_of_colors.append(i.text.split()[0])
        else:
            no_of_colors.append("-")
        #clicking on the next button to go the next page   
        wait = WebDriverWait(driver, 20)
        button = wait.until(EC.visibility_of_element_located((By.XPATH, '//*[@id="main-content"]/div/div[2]/div/div/div/div[2]/div/div/div/div[4]/a/span')))
        driver.execute_script("arguments[0].scrollIntoView();", button)
        button.click()
    except NoSuchElementException:
        break
    except ElementClickInterceptedException:
        break
    except Exception as e:
        pass
        break

In [8]:
len(names), len(category), len(no_of_colors)

(2960, 2960, 2020)

In [138]:
product_details = WebDriverWait(driver, 20).until(EC.presence_of_all_elements_located((By.XPATH,'//a[@class="glass-product-card__assets-link"]')))

for index in range(len(product_details)):
    try:
        product_details = WebDriverWait(driver, 20).until(EC.presence_of_all_elements_located((By.XPATH,'//a[@class="glass-product-card__assets-link"]')))
        # Click on the current product detail
        current_product = product_details[index]
        current_product.click()
        
        # Extract prices
        #price_elements = driver.find_element(By.XPATH, '//div[@class="gl-price-item notranslate"]')
        #price.append(price_elements.text)
        price_element= driver.find_element(By.XPATH,'//*[@id="main-content"]/div[2]/div[2]/div[1]/div[2]/div/div/div/div')
        if price_element:
            #Append  price to the price list
            price.append(price_element.text)
        else:
            #Append sale price to the price list
            sale_price_element = price_elements.find_element(By.XPATH, '//*[@id="main-content"]/div[2]/div[2]/div[1]/div[2]/div/div/div/div[2]')
            price.append(sale_price_element.text)
            

        # Extract color information
        color_elements = driver.find_elements(By.XPATH,'/html/body/div[2]/div/div[1]/div[1]/div/div/div[4]/div[2]/div[2]/div[2]/div[2]/a')
        for i in range(len(color_elements)):
            try:
                color1.append(driver.find_element(By.XPATH,'//div[@class="color-label___2hXaD"]').text)
            except NoSuchElementException:
                color1.append("-")
                i=i+1
            if i<len(color_elements):
                try:
                    color2.append(driver.find_element(By.XPATH,'//div[@class="color-label___2hXaD"]').text)
                except NoSuchElementException:
                    color2.append("-")
                    i += 1
            else:
                color2.append("-")
            if i<len(color_elements):
                try:
                    color3.append(driver.find_element(By.XPATH,'//div[@class="color-label___2hXaD"]').text)
                except NoSuchElementException:
                    color3.append("-")
                    i += 1
            else:
                color3.append("-")
            if i<len(color_elements):
                try:
                    color4.append(driver.find_element(By.XPATH,'//div[@class="color-label___2hXaD"]').text)
                except NoSuchElementException:
                    color4.append("-")
                    i += 1
            else:
                color4.append("-")
            if i<len(color_elements):
                try:
                    color5.append(driver.find_element(By.XPATH,'//div[@class="color-label___2hXaD"]').text)
                except NoSuchElementException:
                    color5.append("-")
                    i += 1
            else:
                color5.append("-")
                
        # Extract sizes and count of sizes
        size_elements = driver.find_elements(By.XPATH,'//*[@id="main-content"]/div[2]/div[2]/section/div[1]/div[2]/button')
        sizes = [size.text for size in size_elements]
        count_of_sizes.append(len(sizes))
        
        #Extract productcode
        driver.execute_script("window.scrollTo(0,1400);")
        details_button = driver.find_element(By.XPATH,'//*[@id="navigation-target-specifications"]/div/button/div[2]')
        if details_button:
            details_button.click()
            try:
                productcode_elements = driver.find_elements(By.XPATH,'//li[@class="gl-vspace-bpall-small"]')
                productcode.append(productcode_elements[5].text.strip("Product code:")[1:])
                
                #Close details button
                close_details_button = driver.find_element(By.XPATH,'//*[@id="navigation-target-specifications"]/div/button/div[2]')
                close_details_button.click()
            except Exception as e:
                print(f'Error Occured while extracting the productcode : {str(e)}')
        else:
            productcode.append("-")
        
        #Extract reviews,ratings
        review_xpath = '//*[@id="navigation-target-reviews"]/div/button/div[2]'
        try:
            reviews_button = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.XPATH, review_xpath)))
            reviews_button.click()
            try:
                reviews_element = WebDriverWait(driver, 5).until(EC.presence_of_element_located((By.XPATH,'//*[@id="navigation-target-reviews"]/div/button/div[1]/div/h2')))
                reviews.append(reviews_element.text.strip("Reviews")[1:])
                
                #extract ratings
                rating_element = driver.find_element(By.XPATH,'//*[@id="navigation-target-reviews"]/div/div/div/div/div/div/div[1]/div/span')
                if rating_element:
                    ratings.append(rating_element.text)
                else:
                    ratings.append('-')
                
                #Extract comfort
                ratings_element = WebDriverWait(driver, 5).until(EC.presence_of_all_elements_located((By.XPATH,'//label[@class="gl-comparison-bar__label gl-comparison-bar__label--active slider-label___29Djz"][1]')))
                if ratings_element:
                    comfort.append(ratings_element[0].text)
                    durability.append(ratings_element[1].text)
                    sizes.append(ratings_element[2].text)
                    
                else:
                    comfort.append("-")
                    durability.append("-")
                    sizes.append("-")

                #Close review button
                close_review_button = driver.find_elements(By.XPATH,'//*[@id="navigation-target-reviews"]/div/button/div[2]')
                close_review_button.click()
            except Exception as e:
                print(f"Error occured while extracting the reviews {str(e)}")
        except NoSuchElementException as e:
            print(f"Review element not found: {str(e)}")
            ratings.append("-")
            reviews.append("-")
            comfort.append("-")
            durability.append("-")
            sizes.append("-")
    
        driver.back()
        wait = WebDriverWait(driver, 20)
        button = wait.until(EC.visibility_of_element_located((By.XPATH, '//*[@id="main-content"]/div/div[2]/div/div/div/div[2]/div/div/div/div[4]/a/span')))
        driver.execute_script("arguments[0].scrollIntoView();", button)
        button.click()
    except Exception as e:
        print(f"Error processing element: {str(e)}")
    

In [159]:
len(reviews),len(ratings),len(comfort),len(durability),len(sizes)

(358, 355, 355, 355, 337)

In [158]:
for _ in range(2,338):
    size.append("-")

In [90]:
for _ in range(108,339):
    productcode.append("-")
for _ in range(332,338):
    count_of_sizes.append("-")
for _ in range(2,338):
    sizes.append("-")

In [94]:
table1 = pd.DataFrame(list(zip(names[:338], no_of_colors[:338], category[:338],price[:338])),
                  columns=['Name', 'No_of_Colors', 'Category', 'Price'])

In [95]:
csv_file_path = 'C:/Users/vaddi/.jupyter/table1.3.csv'

# Write the DataFrame to the CSV file
table1.to_csv(csv_file_path, index=True)

# Display the CSV file path
csv_file_path

'C:/Users/vaddi/.jupyter/table1.3.csv'

In [96]:
table2= pd.DataFrame(list(zip(count_of_sizes[:338], color1[:338],color2[:338],color3[:338],color4[:338],color5[:338],productcode[:338])),columns=['Sizes','color1','color2','color3','color4','color5','product code'])

csv_file_path = 'C:/Users/vaddi/.jupyter/table2.1.csv'
# Write the DataFrame to the CSV file
table2.to_csv(csv_file_path, index=True)

# Display the CSV file path
csv_file_path

'C:/Users/vaddi/.jupyter/table2.1.csv'

In [160]:
table3= pd.DataFrame(list(zip(reviews[:338],size[:338],comfort[:338],durability[:338],ratings[:338])),columns=['Reviews','Size','Comfort','Durability','Ratings'])

csv_file_path = 'C:/Users/vaddi/.jupyter/table3.1.csv'
# Write the DataFrame to the CSV file
table3.to_csv(csv_file_path, index=True)

# Display the CSV file path
csv_file_path

'C:/Users/vaddi/.jupyter/table3.1.csv'

In [161]:
driver.quit()

In [13]:
import selenium
from selenium import webdriver
from bs4 import BeautifulSoup,SoupStrainer
from bs4 import BeautifulSoup
import requests
import pandas as pd
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.common.exceptions import NoSuchElementException
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.action_chains import ActionChains
from selenium.common.exceptions import ElementClickInterceptedException

import re
import time

import warnings
warnings.filterwarnings('ignore')

In [53]:
# sending request to scrap the webpage
url = 'https://us.puma.com/us/en'
page = requests.get(url)
page

<Response [200]>

In [54]:
#Opening the website in the automated window
driver =webdriver.Chrome()
driver.get(url)

In [57]:
#Let's maximize the window
driver.maximize_window()

In [58]:
#Let's locate the search bar in the webpage
sear_in = driver.find_element(By.XPATH,"/html/body/div[1]/div[1]/div/nav/div/div/button[1]")
sear_in.click()

In [59]:
#Sending the input keys to the search bar
in_ = driver.find_element(By.XPATH,"/html/body/div[1]/div[1]/div/nav/div/div/div[1]/div[2]/div[1]/div/form/input")
in_.send_keys("Shoes")

In [61]:
Search = driver.find_element(By.XPATH,'/html/body/div[2]/div[1]/div/nav/div/div/div[1]/div[2]/div[2]/div/div/div/div[1]/div/ul/li[3]/a/span')
Search.click()

In [21]:
#Selecting the gender in the filters to scrap the details gender wise
gender = driver.find_element(By.XPATH,'/html/body/div[2]/div[1]/main/div[1]/section/nav[1]/div/div/div/div[1]/div[2]/div/div[2]/button')
gender.click()

gender1_select = driver.find_element(By.XPATH,'/html/body/div[2]/div[1]/main/div[1]/section/nav[1]/div/div/div/div[1]/div[2]/div/div[2]/div/div/div[2]/div/div/label[1]/div[1]')
gender1_select.click()

In [22]:
close_popup = driver.find_element(By.XPATH,'/html/body/div[2]/div[1]/main/div[1]/section/nav[1]/div/div/div/div[1]/div[2]/div/div[2]/div/div/div[1]/button')
close_popup.click()

In [23]:
#To load all the products in the webpage
for _ in range(20):
    driver.find_element(By.TAG_NAME,'body').send_keys(Keys.END)
    time.sleep(10)

In [24]:
#sending the request to the scrap the page
url = 'https://us.puma.com/us/en/search?q=Shoes&pref_gender=Men&offset=384'
page = requests.get(url)
page

<Response [200]>

In [25]:
soup = BeautifulSoup(page.content,'html.parser')

In [26]:
#Extract the product names
names= [i.text for i in soup.find_all('h3',class_="w-full mobile:text-sm mobile:pr-0 font-bold text-base pr-5 line-clamp-2")]
#Extract the number of colors the product is available in
colors = [i.text.strip('C')[0] for i in soup.find_all('div',class_="relative h-12 mobile:h-10 flex items-center group")]

#extracting the price of each product and the category
prices = []
category = []
for i in soup.find_all('div',class_="flex flex-col flex-none mobile:items-start items-end text-sm md:text-base mobile:mt-2"):
    price_text = i.text
    match = re.search(r'\$\d+.\d+', price_text)
    if match:
        prices.append(match.group())
        category.append("Men")  

In [27]:
#Extracting the reviews
reviews = []
for div in soup.find_all('div', class_="relative w-full flex flex-col gap-2"):
    # Check if the desired class exists within the current div
    class_exists = div.find(class_="ml-1 pl-px text-sm")
    
    # If the class exists, append its text to the result list
    if class_exists:
        reviews.append(class_exists.text.strip("Reviews"))
    else:
        # If the class doesn't exist, append "-"
        reviews.append("-")

In [28]:
#Extracting the ratings
ratings = []
for div in soup.find_all('div', class_="relative w-full flex flex-col gap-2"):
    class_exists = div.find(class_="flex items-center")
    if class_exists:
        ratings_text = class_exists.text
        match = re.search(r'\b\d+\.\d+\b', class_exists.text)
        if match:
            ratings.append(float(match.group()))
        
    else:
        # If the class doesn't exist, append "-"
        ratings.append("-")

In [29]:
len(names),len(prices),len(category),len(reviews),len(ratings),len(colors)


(392, 392, 392, 392, 392, 392)

In [30]:
#Creating the empty lists to extract the product details
count_of_sizes = []
color1= []
color2= []
color3= []
color4 = []
color5 = []
productcode = []
size =[]
comfort = []
durability = []

In [74]:

# Product details elements
product_details = driver.find_elements(By.XPATH, '//div[@class="relative flex mobile:flex-col w-full items-start"]')
# Loop through product details
for index, product in enumerate(product_details):
    try:
        # Click on the current product
        product_details[index].click()
        
        driver.execute_script("window.scrollTo(0, 300);")
        driver.implicitly_wait(1)
        # Extract color information
        color_elements = driver.find_elements(By.XPATH, '/html/body/div[2]/div[1]/main/div[1]/section/div[1]/section[1]/div/div[3]/label/input')

        for i in range(len(color_elements)):
            try:
                color1.append(driver.find_element(By.XPATH, '//h4[@class="tw-6ypl7n tw-lxvy65 tw-6nh9ql tw-p9uz4a tw-1h4nwdw tw-1jgcml9"]').text)
            except NoSuchElementException:
                color1.append("-")
                i = i + 1

            if i < len(color_elements):
                try:
                    color2.append(driver.find_element(By.XPATH, '//h4[@class="tw-6ypl7n tw-lxvy65 tw-6nh9ql tw-p9uz4a tw-1h4nwdw tw-1jgcml9"]').text)
                except NoSuchElementException:
                    color2.append("-")
            else:
                color2.append("-")
                i = i + 1       

            if i < len(color_elements):
                try:
                    color3.append(driver.find_element(By.XPATH, '//h4[@class="tw-6ypl7n tw-lxvy65 tw-6nh9ql tw-p9uz4a tw-1h4nwdw tw-1jgcml9"]').text)
                except NoSuchElementException:
                    color3.append("-")
            else:
                color3.append("-")
                i = i + 1

            if i < len(color_elements):
                try:
                    color4.append(driver.find_element(By.XPATH, '//h4[@class="tw-6ypl7n tw-lxvy65 tw-6nh9ql tw-p9uz4a tw-1h4nwdw tw-1jgcml9"]').text)
                except NoSuchElementException:
                    color4.append("-")
            else:
                color4.append("-")
                i = i + 1

            if i < len(color_elements):
                try:
                    color5.append(driver.find_element(By.XPATH, '//h4[@class="tw-6ypl7n tw-lxvy65 tw-6nh9ql tw-p9uz4a tw-1h4nwdw tw-1jgcml9"]').text)
                except NoSuchElementException:
                    color5.append("-")
            else:
                color5.append("-")

        # Extract sizes and count of sizes
        size_elements = WebDriverWait(driver, 10).until(EC.presence_of_all_elements_located((By.XPATH, '//input[@class="absolute block appearance-none w-full h-full inset-0 cursor-pointer"]')))
        sizes = [size.text for size in size_elements]
        count_of_sizes.append(len(sizes))

        # Scroll down
        driver.execute_script("window.scrollTo(0,1435);")
        driver.implicitly_wait(1)

        # Extract product code
        try:
            productcode_element = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.XPATH, '//*[@id="puma-skip-here"]/div[1]/section/div[1]/section[1]/div/div[9]/ul')))
            productcode.append(productcode_element.text.strip().split(':')[1].strip('\nColor'))
        except Exception as e:
            print(f"error occured while extracting productcode : {str(e)}")
            productcode.append("-")
        
        # Extract quality
        try:
            # Locate and click on the "Read All" button
            read_all_button_elements = WebDriverWait(driver, 20).until(EC.presence_of_element_located((By.XPATH,'//button[@class="group tw-19fr0v0 tw-ozwx86 flex-row after::absolute after::hidden after::pointer-events-none after::-left-1 after::-top-1 after::-right-1 after::-bottom-1 after::border-2 after::border-outline after::rounded-sm tw-10skrdm self-end"][2]')))
            if read_all_button_elements:
                read_all_button_elements.click()
                driver.execute_script("window.scrollTo(0, 1200);")
                ratings_elements = WebDriverWait(driver, 20).until(EC.presence_of_all_elements_located((By.XPATH,'//span[@class="flex space-x-2 font-bold"]')))
                for i, element in enumerate(ratings_elements):
                    if i % 2 == 0:
                        durability.append(element.text.strip("Quality\nAverage :")[0:4])

                    else:
                        comfort.append(element.text.strip("Quality\nAverage :")[0:4])

                # Scroll back to the top
                driver.execute_script("window.scrollTo(0, 0);")

                # Close the review modal
                close_button = driver.find_element(By.XPATH,'//button[@class="group tw-4xsq9o tw-o47a5z"]')
                close_button.click()

        except Exception as e:
            print(f"Error: {str(e)}")
            durability.append("-")
            comfort.append("-")

        # Go back to the product list
        driver.back()
        product_details = WebDriverWait(driver,1).until(EC.presence_of_all_elements_located((By.XPATH, '//div[@class="relative flex mobile:flex-col w-full items-start"]')))
        time.sleep(3)
    except Exception as e:
        print(f"Error: {str(e)}")

In [142]:
driver.execute_script("window.scrollTo(0,0);")

In [62]:
# selecting 'women' from filters, gender to extract the details
gender = driver.find_element(By.XPATH,'/html/body/div[2]/div[1]/main/div[1]/section/nav[1]/div/div/div/div[1]/div[2]/div/div[2]/button')
gender.click()

gender1_select = driver.find_element(By.XPATH,'/html/body/div[2]/div[1]/main/div[1]/section/nav[1]/div/div/div/div[1]/div[2]/div/div[2]/div/div/div[2]/div/div/label[1]/div[1]')
gender1_select.click()

In [63]:
gender2_select = driver.find_element(By.XPATH,'/html/body/div[2]/div[1]/main/div[1]/section/nav[1]/div/div/div/div[1]/div[2]/div/div[2]/div/div/div[2]/div/div/label[2]/div[1]')
gender2_select.click()

close_popup = driver.find_element(By.XPATH,'/html/body/div[2]/div[1]/main/div[1]/section/nav[1]/div/div/div/div[1]/div[2]/div/div[2]/div/div/div[1]/button')
close_popup.click()

In [64]:
#Loading all the products in the webpage
for _ in range(20):
    driver.find_element(By.TAG_NAME,'body').send_keys(Keys.END)
    time.sleep(10)

In [65]:
url = 'https://us.puma.com/us/en/search?q=Shoes&pref_gender=Women&offset=312'
page = requests.get(url)
page

soup = BeautifulSoup(page.content,'html.parser')

In [66]:
#Extract the names of the products
for i in soup.find_all('h3',class_="w-full mobile:text-sm mobile:pr-0 font-bold text-base pr-5 line-clamp-2"):
    names.append(i.text)
    category.append("Women")  
#Extract the  price
for i in soup.find_all('div',class_="flex flex-col flex-none mobile:items-start items-end text-sm md:text-base mobile:mt-2"):
    price_text = i.text
    match = re.search(r'\$\d+.\d+', price_text)
    if match:
        prices.append(match.group())
#Extract the no_of_colors
for i in soup.find_all('div',class_="relative h-12 mobile:h-10 flex items-center group"):
    colors.append(i.text.strip('C')[0]) 
#Extracting the reviews
for div in soup.find_all('div', class_="relative w-full flex flex-col gap-2"):
    # Check if the desired class exists within the current div
    class_exists = div.find(class_="ml-1 pl-px text-sm")
    # If the class exists, append its text to the result list
    if class_exists:
        reviews.append(class_exists.text.strip("Reviews"))
    else:
        # If the class doesn't exist, append "-"
        reviews.append("-")
#Extracting the ratings
for div in soup.find_all('div', class_="relative w-full flex flex-col gap-2"):
    class_exists = div.find(class_="flex items-center")
    if class_exists:
        ratings_text = class_exists.text
        match = re.search(r'\b\d+\.\d+\b', class_exists.text)
        if match:
            ratings.append(float(match.group()))      
    else:
        # If the class doesn't exist, append "-"
        ratings.append("-")

In [79]:
# Product details elements
product_details = driver.find_elements(By.XPATH, '//div[@class="relative flex mobile:flex-col w-full items-start"]')
# Loop through product details
for index, product in enumerate(product_details):
    try:
        # Click on the current product
        product_details[index].click()
        
        driver.execute_script("window.scrollTo(0, 300);")
        driver.implicitly_wait(1)
        # Extract color information
        color_elements = driver.find_elements(By.XPATH, '/html/body/div[2]/div[1]/main/div[1]/section/div[1]/section[1]/div/div[3]/label/input')

        for i in range(len(color_elements)):
            try:
                color1.append(driver.find_element(By.XPATH, '//h4[@class="tw-6ypl7n tw-lxvy65 tw-6nh9ql tw-p9uz4a tw-1h4nwdw tw-1jgcml9"]').text)
            except NoSuchElementException:
                color1.append("-")
                i = i + 1

            if i < len(color_elements):
                try:
                    color2.append(driver.find_element(By.XPATH, '//h4[@class="tw-6ypl7n tw-lxvy65 tw-6nh9ql tw-p9uz4a tw-1h4nwdw tw-1jgcml9"]').text)
                except NoSuchElementException:
                    color2.append("-")
            else:
                color2.append("-")
                i = i + 1       

            if i < len(color_elements):
                try:
                    color3.append(driver.find_element(By.XPATH, '//h4[@class="tw-6ypl7n tw-lxvy65 tw-6nh9ql tw-p9uz4a tw-1h4nwdw tw-1jgcml9"]').text)
                except NoSuchElementException:
                    color3.append("-")
            else:
                color3.append("-")
                i = i + 1

            if i < len(color_elements):
                try:
                    color4.append(driver.find_element(By.XPATH, '//h4[@class="tw-6ypl7n tw-lxvy65 tw-6nh9ql tw-p9uz4a tw-1h4nwdw tw-1jgcml9"]').text)
                except NoSuchElementException:
                    color4.append("-")
            else:
                color4.append("-")
                i = i + 1

            if i < len(color_elements):
                try:
                    color5.append(driver.find_element(By.XPATH, '//h4[@class="tw-6ypl7n tw-lxvy65 tw-6nh9ql tw-p9uz4a tw-1h4nwdw tw-1jgcml9"]').text)
                except NoSuchElementException:
                    color5.append("-")
            else:
                color5.append("-")

        # Extract sizes and count of sizes
        size_elements = WebDriverWait(driver, 10).until(EC.presence_of_all_elements_located((By.XPATH, '//input[@class="absolute block appearance-none w-full h-full inset-0 cursor-pointer"]')))
        sizes = [size.text for size in size_elements]
        count_of_sizes.append(len(sizes))

        # Scroll down
        driver.execute_script("window.scrollTo(0,1435);")
        driver.implicitly_wait(1)

        # Extract product code
        try:
            productcode_element = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.XPATH, '//*[@id="puma-skip-here"]/div[1]/section/div[1]/section[1]/div/div[9]/ul')))
            productcode.append(productcode_element.text.strip().split(':')[1].strip('\nColor'))
        except Exception as e:
            print(f"error occured while extracting productcode : {str(e)}")
            productcode.append("-")
        
        # Extract quality
        try:
            # Locate and click on the "Read All" button
            read_all_button_elements = WebDriverWait(driver, 20).until(EC.presence_of_element_located((By.XPATH,'//button[@class="group tw-19fr0v0 tw-ozwx86 flex-row after::absolute after::hidden after::pointer-events-none after::-left-1 after::-top-1 after::-right-1 after::-bottom-1 after::border-2 after::border-outline after::rounded-sm tw-10skrdm self-end"][2]')))
            if read_all_button_elements:
                read_all_button_elements.click()
                driver.execute_script("window.scrollTo(0, 1200);")
                ratings_elements = WebDriverWait(driver, 20).until(EC.presence_of_all_elements_located((By.XPATH,'//span[@class="flex space-x-2 font-bold"]')))
                for i, element in enumerate(ratings_elements):
                    if i % 2 == 0:
                        durability.append(element.text.strip("Quality\nAverage :")[0:4])

                    else:
                        comfort.append(element.text.strip("Quality\nAverage :")[0:4])

                # Scroll back to the top
                driver.execute_script("window.scrollTo(0, 0);")

                # Close the review modal
                close_button = driver.find_element(By.XPATH,'//button[@class="group tw-4xsq9o tw-o47a5z"]')
                close_button.click()

        except Exception as e:
            print(f"Error: {str(e)}")
            durability.append("-")
            comfort.append("-")

        # Go back to the product list
        driver.back()
        product_details = WebDriverWait(driver,1).until(EC.presence_of_all_elements_located((By.XPATH, '//div[@class="relative flex mobile:flex-col w-full items-start"]')))
        time.sleep(3)
    except Exception as e:
        print(f"Error: {str(e)}")

In [170]:
len(names),len(ratings),len(reviews),len(prices),len(category),len(colors)

(707, 707, 707, 707, 707, 707)

In [171]:
#creating table one
table1 = pd.DataFrame(list(zip(names, colors, category,prices)),
                  columns=['Name', 'No_of_Colors', 'Category', 'Price'])

In [172]:
csv_file_path = 'C:/Users/vaddi/.jupyter/table1.2.csv'

# Write the DataFrame to the CSV file
table1.to_csv(csv_file_path, index=True)

# Display the CSV file path
csv_file_path

'C:/Users/vaddi/.jupyter/table1.2.csv'

In [76]:
table2= pd.DataFrame(list(zip(count_of_sizes[:200], color1[:200],color2[:200],color3[:200],color4[:200],color5[:200],productcode[:200])),columns=['Sizes','color1','color2','color3','color4','color5','product code'])

csv_file_path = 'C:/Users/vaddi/.jupyter/table2.2.csv'
# Write the DataFrame to the CSV file
table2.to_csv(csv_file_path, index=True)

# Display the CSV file path
csv_file_path

'C:/Users/vaddi/.jupyter/table2.2.csv'

In [85]:
for _ in range(1,200):
    size.append("-")

In [86]:
table3= pd.DataFrame(list(zip(reviews[:200],size[:200],comfort[:200],durability[:200],ratings[:200])),columns=['Reviews','Size','Comfort','Durability','Ratings'])

csv_file_path = 'C:/Users/vaddi/.jupyter/table3.2.csv'
# Write the DataFrame to the CSV file
table3.to_csv(csv_file_path, index=True)

# Display the CSV file path
csv_file_path

'C:/Users/vaddi/.jupyter/table3.2.csv'

In [84]:
len(reviews), len(size), len(comfort), len(durability), len(ratings)

(707, 198, 858, 881, 707)

In [73]:
#driver.execute_script('return document.body.scrollHeight')
#driver.execute_script("window.scrollTo(0,1200);")
#driver.execute_script("window.scrollTo(0,2000);")

In [78]:
driver.quit()

In [48]:
import selenium
from selenium import webdriver
from bs4 import BeautifulSoup,SoupStrainer
import requests
import pandas as pd
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.common.exceptions import NoSuchElementException
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.action_chains import ActionChains
from selenium.common.exceptions import ElementClickInterceptedException
from selenium.webdriver.support.ui import WebDriverWait
import re
import time

import warnings
warnings.filterwarnings('ignore')

In [95]:
#Sending requests to scrap the website
url = 'https://www.skechers.com'
page = requests.get(url)
page

<Response [200]>

In [96]:
#opening the website inthe automated browser
driver =webdriver.Chrome()
driver.get(url)

In [97]:
#Maximizing the window
driver.maximize_window()

In [98]:
#sending the input keys
search_input = driver.find_element(By.XPATH,'/html/body/div[2]/header/nav[2]/div/div[1]/div/div/div[2]/div[1]/div/div/form/input[1]')
search_input.send_keys("Shoes")

In [101]:
search = driver.find_element(By.XPATH,'//*[@id="rfk_search_container"]/div[2]/div/ul/li/ul/li[8]/a/span')
search.click()

In [102]:
#Verifying the requests status again
url= 'https://www.skechers.com/search/?q=shoe'
page = requests.get(url)
page

<Response [403]>

In [103]:
headers = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"}

In [104]:
r= requests.get("https://www.skechers.com/search/?q=shoe", headers = headers)
r

<Response [403]>

Cannot scrap the website as the requests status code is 403

In [105]:
driver.quit()

In [106]:
import selenium
from selenium import webdriver
from bs4 import BeautifulSoup,SoupStrainer
import requests
import pandas as pd
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.common.exceptions import NoSuchElementException
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.action_chains import ActionChains
from selenium.common.exceptions import ElementClickInterceptedException
from selenium.webdriver.support.ui import WebDriverWait
import re
import time

import warnings
warnings.filterwarnings('ignore')

In [107]:
url= 'https://www.columbia.com/'
page = requests.get(url)
page

<Response [200]>

In [108]:
driver =webdriver.Chrome()
driver.get(url)

In [109]:
driver.maximize_window()

In [110]:
input_ = driver.find_element(By.XPATH,'/html/body/header/nav/div/div/div/div/div[2]/div/nav/ul/li[4]/a/span[1]')
input_.click()

"Access to this page has been denied because we believe you are using automation tools to browse the website." ontrying to run the code

Hence cannot scrap the website

In [114]:
driver.quit()

In [115]:
import selenium
from selenium import webdriver
from bs4 import BeautifulSoup,SoupStrainer
import requests
import pandas as pd
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.common.exceptions import NoSuchElementException
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.action_chains import ActionChains
from selenium.common.exceptions import ElementClickInterceptedException
from selenium.webdriver.support.ui import WebDriverWait
import re
import time

import warnings
warnings.filterwarnings('ignore')

In [116]:
url = 'https://www.woodlandworldwide.com/'
page = requests.get(url)
page

<Response [200]>

In [117]:
driver = webdriver.Chrome()
driver.get(url)

In [118]:
driver.maximize_window()

In [120]:
button = driver.find_element(By.XPATH,'/html/body/div[1]/div[2]/section[2]/div/div/div[2]/div/div[2]/button')
button.click()

In [121]:
for i in driver.find_elements(By.CLASS_NAME,"productcardplp_productname__F8lks"):
    names.append(i.text)
    category.append("men")

In [122]:
for i in driver.find_elements(By.CLASS_NAME,"productcardplp_productprice__jdlz6"):
    price.append(i.text.strip().split("₹")[1])

In [123]:
product_details = driver.find_elements(By.XPATH,'/html/body/div/div/div/div[4]/div/div/div/div/div/a/div/div[1]/div[5]/div/div/span/img')

for index in range(len(product_details)):
    try:
        # Re-fetch the product details on each iteration
        product_details = driver.find_elements(By.XPATH,'/html/body/div/div/div/div[4]/div/div/div/div/div/a/div/div[1]/div[5]/div/div/span/img')
        # Click on the current product detail
        current_product = product_details[index]
        current_product.click()
        
        # Extract colors and count of colors
        colors_elements = driver.find_elements(By.CLASS_NAME, "colorpicker_option__B8oxU ")
        colors = [color.text for color in colors_elements]
        no_of_colors.append(len(colors))

        # Extract sizes and count of sizes
        size_elements = driver.find_elements(By.CLASS_NAME, "sizepicker_sizecontainer__xrprC")
        sizes = [size.text for size in size_elements]
        count_of_sizes.append(len(sizes))
        
        ActionChains(driver).send_keys(Keys.ARROW_DOWN).perform()
        # Extract product code
        productcode_element = driver.find_element(By.XPATH, '/html/body/div[1]/div[1]/main/div/section/div[1]/section[1]/div/div[9]/ul/li[1]')
        productcode.append(productcode_element.text.strip().split(':')[1].strip('C'))


        # Extract reviews
        reviews_element = driver.find_element(By.CLASS_NAME,'/html/body/div[1]/div[1]/main/div/section/div[1]/section[1]/div/div[11]/div[1]/h2')
        reviews.append(reviews_element.text.split('(')[1].strip(')'))

        # Extract ratings
        ratings_element = driver.find_element(By.XPATH, '//a[@href="#product-reviews"]')
        raw_rating = ratings_element.text.strip(" Average")
        rating = raw_rating.split(":")[0].strip("/")
        ratings.append(rating)
        #ratings.append(ratings_element.text.strip("S"))

        # Extract color information
        # Extract color information
        color_elements = driver.find_elements(By.XPATH,'//input[@class="absolute appearance-none block w-full h-full inset-0 cursor-pointer"]')
        for i in range(5):
            if i < len(color_elements):
                color_text = color_elements[i].get_attribute("value")
                color1.append(color_text if i == 0 else "-")
                color2.append(color_text if i == 1 else "-")
                color3.append(color_text if i == 2 else "-")
                color4.append(color_text if i == 3 else "-")
                color5.append(color_text if i == 4 else "-")
            else:
                # If there are fewer than 5 colors, append "-"
                color1.append("-")
                color2.append("-")
                color3.append("-")
                color4.append("-")
                color5.append("-")


        # Navigate back to the product list page
        driver.back()

    except Exception as e:
        print(f"Error processing element: {str(e)}")


Error processing element: Message: element not interactable
  (Session info: chrome=120.0.6099.72)
Stacktrace:
	GetHandleVerifier [0x00007FF7BE562142+3514994]
	(No symbol) [0x00007FF7BE180CE2]
	(No symbol) [0x00007FF7BE0274C3]
	(No symbol) [0x00007FF7BE072D29]
	(No symbol) [0x00007FF7BE066A0F]
	(No symbol) [0x00007FF7BE095FEA]
	(No symbol) [0x00007FF7BE0663B6]
	(No symbol) [0x00007FF7BE096490]
	(No symbol) [0x00007FF7BE0B28F6]
	(No symbol) [0x00007FF7BE095D93]
	(No symbol) [0x00007FF7BE064BDC]
	(No symbol) [0x00007FF7BE065C64]
	GetHandleVerifier [0x00007FF7BE58E16B+3695259]
	GetHandleVerifier [0x00007FF7BE5E6737+4057191]
	GetHandleVerifier [0x00007FF7BE5DE4E3+4023827]
	GetHandleVerifier [0x00007FF7BE2B04F9+689705]
	(No symbol) [0x00007FF7BE18C048]
	(No symbol) [0x00007FF7BE188044]
	(No symbol) [0x00007FF7BE1881C9]
	(No symbol) [0x00007FF7BE1788C4]
	BaseThreadInitThunk [0x00007FFC84A0257D+29]
	RtlUserThreadStart [0x00007FFC84F8AA58+40]

Error processing element: Message: element not int

# Combining the similar data into one table

# Table1

In [17]:
file_paths = [
    'C:\\Users\\vaddi\\OneDrive\\Desktop\\assignment\\table1.csv',
    'C:\\Users\\vaddi\\OneDrive\\Desktop\\assignment\\table1.2.csv',
    'C:\\Users\\vaddi\\OneDrive\\Desktop\\assignment\\table1.3.csv'
]

# Read each CSV file into a DataFrame
dfs = [pd.read_csv(file) for file in file_paths]

# Concatenate the DataFrames vertically
combined_df = pd.concat(dfs, ignore_index=True)


In [21]:
csv_file_path = 'C:\\Users\\vaddi\\OneDrive\\Desktop\\assignment\\TABLE1.csv'
combined_df.to_csv(csv_file_path, index=True)

# Table2

In [30]:
file_paths = [
    'C:/Users/vaddi/OneDrive/Desktop/assignment/table2.csv',
    'C:/Users/vaddi/OneDrive/Desktop/assignment/table2.1.csv',
    'C:/Users/vaddi/OneDrive/Desktop/assignment/table2.2.csv'
]

# Read each CSV file into a DataFrame
dfs = [pd.read_csv(file) for file in file_paths]

# Concatenate the DataFrames vertically
combined_df = pd.concat(dfs, ignore_index=True)

In [31]:
csv_file_path = 'C:\\Users\\vaddi\\OneDrive\\Desktop\\assignment\\TABLE2.csv'
combined_df.to_csv(csv_file_path, index=True)

In [32]:
combined_df.shape

(1161, 8)

# Table3

In [33]:
file_paths = [
    'C:/Users/vaddi/OneDrive/Desktop/assignment/table3.csv',
    'C:/Users/vaddi/OneDrive/Desktop/assignment/table3.1.csv',
    'C:/Users/vaddi/OneDrive/Desktop/assignment/table3.2.csv'
]

# Read each CSV file into a DataFrame
dfs = [pd.read_csv(file) for file in file_paths]

# Concatenate the DataFrames vertically
combined_df = pd.concat(dfs, ignore_index=True)

In [34]:
csv_file_path = 'C:\\Users\\vaddi\\OneDrive\\Desktop\\assignment\\TABLE3.csv'
combined_df.to_csv(csv_file_path, index=True)